### PR TITLE
feat(eslint-plugin): add "enum-const-style" rule

### LINKED ISSUE #280: Rule proposal: Const enum
The user would be able to configure this rule to either enforce always using `const enum` or always disallow it.

Relevant TSLint issue: https://github.com/palantir/tslint/issues/1798

### PR DESCRIPTION
**Pull request by mohsen1**
Closes #280

### CODE REVIEW

**Comment by mysticatea** [comment]
Thank you for your contribution!

I'd like to separate rules, and in order to separate, would you change the rule name as narrowing?

> https://github.com/typescript-eslint/typescript-eslint/issues/280#issuecomment-464431173

**Comment by mysticatea** [comment]
Thank you for update!

Code and tests look good to me.
It will be better if the document described the advantage for both settings. Currently, I feel unfair because the document doesn't describe the pros of `const enum`.

And would you reorder the document content a bit?

```md
# title (rule ID)

Description of context why you use this rule.

## Rule Details

Description of what the rule does.

## Options

Description of options.
```

**Comment by mohsen1** [comment]
I don't get any coverage errors locally. How can I re-run it?

**Comment by armano2** [comment]
depends on editor, in Intelij Idea i'm just pressing 
image

for vscode you will need plugin https://github.com/jest-community/vscode-jest#how-do-i-show-code-coverage

**Comment by bradzacher** [comment]
if you cd into the plugin folder (rather than just in the project root), then run `yarn test` it should run the tests for just the plugin, and then dump the coverage to the console.

File: packages/eslint-plugin/src/rules/enum-const-style.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/**
```
Comment (JamesHenry) [inline packages/eslint-plugin/src/rules/enum-const-style.ts]: Sorry we've changed to using all-contributors now and removed all other occurrences of these file-level author comments, please could you remove this

**Comment by mohsen1** [comment]
I don't see the failure report in codecov. Can you point me to what's actually not passing?

**Comment by bradzacher** [comment]
@mohsen1 https://github.com/typescript-eslint/typescript-eslint/pull/290#issuecomment-464565780

We have a minimum target of 90% coverage on the files that a PR changes.
In particular you only have 87.5% coverage over your new file, which means that you need to look at your branches and make sure you've covered them with your tests.

It's because your file is so short, so one line really messes up the total %age.
Add a valid test case for the `'never`' option?

**Comment by bradzacher** [comment]
One piece of feedback for you with your PRs - don't use the ammend + force push workflow when submitting PRs.

Doing so makes it hard for us to review because git/github doesn't track the old versions of the same commit, which means that we can't compare your new changes to the previous changes.

i.e. we have to re-review the entire PR each time you push, because we don't know what has changed.

**Comment by JamesHenry** [comment]
@mohsen1 when do you think you’ll have chance to review the feedback?

**Comment by JamesHenry** [comment]
Hi @mohsen1, just following up on this again as it's been a few months since the last comment

File: packages/eslint-plugin/docs/rules/enum-const-style.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+```
+
+### `always` config
+
+Only const enums are allowed
+
+Examples of **incorrect** code for this rule with `always` config:
+
+```ts
+enum const Foo {
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/enum-const-style.md]:
```suggestion
const enum Foo {
```

File: packages/eslint-plugin/docs/rules/enum-const-style.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+```
+
+Examples of **correct** code for this rule with `always` config:
+
+```ts
+const Foo {
+  ONE,
+  TWO,
+}
+```
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/enum-const-style.md]: I think the examples are the wrong way around

Comment (armano2) [reply packages/eslint-plugin/docs/rules/enum-const-style.md]: actually i haven't read docs, but i did small refactoring in rule source code

**Pull request closed without merging**